README.md
# dde_dockswitch
### List switcher for Deepin Dock

This is an implementation of [Launcher List Indicator](https://github.com/SergKolo/launcher-list-indicator) for Deepin Desktop Environment. 

If you like this indicator, please consider donating to support the developer
[![paypalbutton](https://www.paypal.com/en_US/i/btn/btn_donate_LG.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=CB9L72S9LEF66)



debian/changelog
dde-dockswitch (0.1-1) unstable; urgency=low

  * source package automatically created by stdeb 0.8.5

 -- Sergiy Kolodyazhnyy <1047481448@qq.com>  Sat, 09 Jun 2018 05:31:36 +0800


dde_dockswitch/dialogs.py
import os
from gi.repository import Gtk
import subprocess
import config_ctrl

def show_about(*args):
    print(args)
    builder = Gtk.Builder()
    builder.add_from_file(os.path.join( 
            os.path.dirname(__file__),
            'dde_dockswitch_about.glade'  ) )
    #print([ i.get_children() for i in builder.get_objects() ])

    # might be better to get children of the 
    # ButtonBox and actually connect Close button
    # to a signal handler
    
    w = builder.get_object("about_dialog")
    response = w.run()
    print("Response",response)
    w.destroy()


def remove_entry_diag(*args):

    dialog = [
        'zenity', '--list', '--checklist', '--title', 
        "Select lists to remove", '--column', "", '--column', "Lists",
    ] 
    lists = config_ctrl.read_config_file()
    
    for i in  lists.keys():
        dialog.append('FALSE')
        dialog.append(i)

    out = subprocess.check_output(dialog).decode().strip()

    for list_name in out.split('|'):
        lists.pop(list_name)

    config_ctrl.write_config_file(lists) 


dde_dockswitch/dock_ctrl.py
# functions responsible for Deepin dock control
import dbus_ctrl

dock_obj = [ "session", "com.deepin.dde.daemon.Dock", 
                "/com/deepin/dde/daemon/Dock", "com.deepin.dde.daemon.Dock"]

def get_desk_files(onlydocked=True):
    """ Obtains .desktop files of entries currently on the dock.
        Defaults to returning only those that are pinned to dock.
        If onlydocked=False, all entries on the dock (including 
        running but not pinned entries) get returned"""
    desk_files = []
    global dock_obj    
    for i in dbus_ctrl.get_dbus_property(*dock_obj,"Entries"):
        entry_obj = [ "session","com.deepin.dde.daemon.Dock", i, 
                      "com.deepin.dde.daemon.Dock.Entry", "DesktopFile"]
        desktop_file = dbus_ctrl.get_dbus_property(*entry_obj)
        if onlydocked and not dbus_ctrl.call_dbus_method(*dock_obj,"IsDocked",[desktop_file]):
            continue
        desktop_file = dbus_ctrl.get_dbus_property(*entry_obj)
        desk_files.append(str(desktop_file))
    return desk_files

def rotate_dock(*args):
    desk_files = get_desk_files(onlydocked=False)
    fill_dock(None,reversed(desk_files))

def clear_dock(*args):
    """  Undocks all entries. Running appications will still 
         appear but won't be pinned"""
    global dock_obj
    for desk_file in get_desk_files():
        dbus_ctrl.call_dbus_method(*dock_obj,"RequestUndock",[desk_file])

def fill_dock(ignore,desk_files,append=False):
    """ Fills the dock with Entries as provided by desk_files list,
        in the order they appear on the list"""
    global dock_obj
    if not append:
        clear_dock()
    for index,desk_file in enumerate(desk_files):
        dbus_ctrl.call_dbus_method(*dock_obj,"RequestDock",[desk_file,index])


dde_dockswitch/menu_builder.py
from gi.repository import Gtk
from collections import OrderedDict
import os,json
import dbus_ctrl
import dialogs

def add_menu_item(menu_obj, type=Gtk.MenuItem,
                  icon=None, label="HelloWorld", action=None, args=[]):
    """ dynamic function that can add menu items depending on
        the item type and other arguments"""

    menu_item = None
    if type is Gtk.ImageMenuItem and label:
        menu_item = Gtk.ImageMenuItem.new_with_label(label)
        menu_item.set_always_show_image(True)
        if '/' in icon:
            icon = Gtk.Image.new_from_file(icon)
        else:
            icon = Gtk.Image.new_from_icon_name(icon, 48)
        menu_item.set_image(icon)
    elif type is Gtk.ImageMenuItem and not label:
        menu_item = Gtk.ImageMenuItem()
        menu_item.set_always_show_image(True)
        if '/' in icon:
            icon = Gtk.Image.new_from_file(icon)
        else:
            icon = Gtk.Image.new_from_icon_name(icon, 16)
        menu_item.set_image(icon)
    elif type is Gtk.MenuItem:
        menu_item = Gtk.MenuItem(label)
    elif type is Gtk.SeparatorMenuItem:
        menu_item = Gtk.SeparatorMenuItem()
    if action:
        menu_item.connect('activate', action, *args)
    menu_obj.append(menu_item)
    menu_obj.show()

def add_submenu(top_menu, label="HelloWorld",**kwargs):
    """ utility function for adding submenus"""
    menuitem = Gtk.MenuItem(label)
    if kwargs and 'icon' in kwargs.keys():
        menuitem = Gtk.ImageMenuItem.new_with_label(label)
        menuitem.set_always_show_image(True)
        if '/' in kwargs['icon']:
            icon = Gtk.Image.new_from_file(kwargs['icon'])
        else:
            icon = Gtk.Image.new_from_icon_name(kwargs['icon'], 48)
        menuitem.set_image(icon)
    submenu = Gtk.Menu()
    menuitem.set_submenu(submenu)
    top_menu.append(menuitem)
    menuitem.show()
    return submenu

def build_base_menu(menu_obj):
    controls = add_submenu(menu_obj,label="Indicator Controls")
    base_menu=OrderedDict([
        ("About", {
            "icon": "info",
            "type": Gtk.ImageMenuItem,
            "action": dialogs.show_about
        }),
        ("Exit",{
            "icon": "exit",
            "type": Gtk.ImageMenuItem,
            "action": Gtk.main_quit
        })
    ])

    for key,val in base_menu.items():
        add_menu_item(controls,label=key,**val,args=[])


dde_dockswitch/dde_dockswitch.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# Copyright © 2018 Sergiy Kolodyazhnyy <1047481448@qq.com>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import gi,os,json
gi.require_version('AppIndicator3', '0.1')
from gi.repository import GLib as glib
from gi.repository import AppIndicator3 
from gi.repository import Gtk
from gi.repository import Gdk
from gi.repository import Gio
from gi.repository.Gdk import ScrollDirection
from collections import OrderedDict

import subprocess

import menu_builder, dialogs,config_ctrl,dbus_ctrl,dock_ctrl


class DeepinDockSwitch(object):

    all_lists = OrderedDict()

    def __init__(self):
        self.app = AppIndicator3.Indicator.new(
            'dde_dock_list', "",
            AppIndicator3.IndicatorCategory.OTHER
        )

        config_file = os.path.join(os.environ["HOME"],".config/dde_dockswitch/docklists.json")

        self.app.set_status(AppIndicator3.IndicatorStatus.ACTIVE)
        self.make_menu()

        icon = '/usr/share/pixmaps/dde_dockswitch_icon.png'
        if os.path.exists(icon):
            self.app.set_icon(icon)
        else:
            icon = os.path.join(os.path.dirname(__file__),'dde_dockswitch_icon.png')
            self.app.set_icon(icon)

    def run(self):
        """ Launches the indicator """
        try:
            Gtk.main()
        except KeyboardInterrupt:
            pass

    def update():
        pass

    def callback():
        pass


    def make_menu(self, *args):
        """ generates entries in the indicator"""
        global all_lists
        if hasattr(self, 'app_menu'):
            for item in self.app_menu.get_children():
                self.app_menu.remove(item)
        self.app_menu = Gtk.Menu()

        # Wrappers for menu entries that call refresh on main menu

        def fill_dock_and_update(ignore,fill):
             dock_ctrl.fill_dock(None,fill)
             self.make_menu()

        def remove_entry_wrap(ignore):
            dialogs.remove_entry_diag()
            self.make_menu()

        all_lists = config_ctrl.read_config_file()
        for list_label,desk_files in all_lists.items():
            if desk_files == list(dock_ctrl.get_desk_files()):
                list_label = "\u2605 " + list_label 
            item_params={ "label": list_label, 
                          "action": fill_dock_and_update,
                          "args": [desk_files]
            }
            menu_builder.add_menu_item(self.app_menu,**item_params)

        menu_builder.add_menu_item(self.app_menu,type=Gtk.SeparatorMenuItem)

        dock_controls = [
            { "label": "Save", 
              "type": Gtk.ImageMenuItem,
              "icon": "gtk-save", 
              "action": self.record_currently_docked,
              "args":[] },

            { "label": "Rotate",
              "type": Gtk.ImageMenuItem,
              "icon": "object-flip-vertical",
              "action": dock_ctrl.rotate_dock,
              "args": []},

            { "label": "Remove",
              "type": Gtk.ImageMenuItem,
              "icon": "list-remove",
              "action": remove_entry_wrap,
              "args": []},

            { "label": "Clear",
              "type": Gtk.ImageMenuItem,
              "icon": "user-trash",
              "action": dock_ctrl.clear_dock,
              "args": []}
        ]

        dock_ctrl_submenu = menu_builder.add_submenu(self.app_menu,label="Dock Controls")
        for i in dock_controls:
            menu_builder.add_menu_item(dock_ctrl_submenu,**i)

        menu_builder.build_base_menu(self.app_menu)
        self.app_menu.show_all()
        self.app.set_menu(self.app_menu)

    def record_currently_docked(self,*args):
        global all_lists
        docked_desk_files = dock_ctrl.get_desk_files(onlydocked=True)
        name = self.run_cmd(['zenity','--entry', 
                      '--text',"Name this list",
                      '--entry-text',"Default entry"]).decode().rstrip()
        if name:
            # this probably will have to be global
            all_lists = {**all_lists,name: docked_desk_files}
            config_ctrl.write_config_file(all_lists)
            self.make_menu()
        


    def run_cmd(self, cmdlist):
        """ utility: reusable function for running external commands """
        try:
            stdout = subprocess.check_output(cmdlist)
        except subprocess.CalledProcessError:
            # TODO
            pass
        else:
            if stdout:
                return stdout


        

if __name__ == '__main__':
    switch = DeepinDockSwitch()
    switch.run()
